powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/persistence/BasePersistenceStore.java

```diff
@@ -221,8 +235,10 @@ private String getHashedIdempotencyKey(JsonNode data) {
         if (isMissingIdemPotencyKey(node)) {
             if (throwOnNoIdempotencyKey) {
                 throw new IdempotencyKeyException("No data found to create a hashed idempotency key");
+            } else {
+                LOG.warn("No data found to create a hashed idempotency key. JMESPath: {}", eventKeyJMESPath);
+                throw new NullIdempotencyKeyException("No data found to create a hashed idempotency key");
```

Comment:
Should this method return an Optional instead of throw an Exception? It would save us raising an Exception and handling in each place.

Reply:
I replaced with Optional, we don't throw an exception anymore but still need to handle the Optional in each place... but that's OK.